cdap-source-control/src/main/java/io/cdap/cdap/sourcecontrol/operationrunner/InMemorySourceControlOperationRunner.java

```diff
@@ -67,11 +71,46 @@ public PushAppResponse push(NamespaceId namespace, RepositoryConfig repoConfig,
       //TODO: CDAP-20371, Add retry logic here in case the head at remote moved while we are doing push
       return writeAppDetailAndPush(repositoryManager, appToPush, commitDetails);
     } finally {
+      closeRepositoryMangerQuietly(repositoryManager);
+    }
+  }
+
+  @Override
+  public PullAppResponse pull(ApplicationReference appRef, RepositoryConfig repoConfig) throws PullFailureException,
+    NotFoundException, AuthenticationConfigException {
+    RepositoryManager repositoryManager = repoManagerFactory.create(appRef.getNamespaceId(), repoConfig);
+    String applicationName = appRef.getApplication();
+    String configFileName = generateConfigFileName(applicationName);
+    Path appRelativePath = repositoryManager.getFileRelativePath(configFileName);
+    LOG.info("Cloning remote to pull application {}", applicationName);
+    try {
+      String commitID = repositoryManager.cloneRemote();
+      Path filePathToRead;
       try {
-        repositoryManager.close();
-      } catch (IOException e) {
-        LOG.warn("Failed to close the RepositoryManager, there may be leftover files", e);
+        filePathToRead = validateAppConfigRelativePath(repositoryManager, appRelativePath);
+      } catch (IllegalArgumentException e) {
+        throw new PullFailureException(String.format("Failed to pull application %s: %s",
+                                                     applicationName,
+                                                     e.getMessage()), e);
       }
+      if (!Files.exists(filePathToRead)) {
+        throw new NotFoundException(String.format("App with name %s not found in git repository", applicationName));
+      }
+      String contents = new String(Files.readAllBytes(filePathToRead), StandardCharsets.UTF_8);
```

Comment:
can we deserialize it here instead of passing it around as a string?

Reply:
Done.